diff --git a/src/race.bash b/src/race.bash
index 3ce91a345f..77fa216613 100755
--- a/src/race.bash
+++ b/src/race.bash
@@ -33,5 +33,5 @@ if [ ! -f make.bash ]; then
 	exit 1
 fi
 . ./make.bash --no-banner
-go install -race std
-go tool dist test -race
+#go install -race std
+#go tool dist test -race
diff --git a/src/run.bash b/src/run.bash
index cb474325c4..e7094aaa3c 100755
--- a/src/run.bash
+++ b/src/run.bash
@@ -50,4 +50,4 @@ if ulimit -T &> /dev/null; then
 fi
 
 export GOPATH=/nonexist-gopath
-exec ../bin/go tool dist test -rebuild "$@"
+# exec ../bin/go tool dist test -rebuild "$@"
diff --git a/src/runtime/lock_wasip1.go b/src/runtime/lock_wasip1.go
index 55153c3a05..55ffab233d 100644
--- a/src/runtime/lock_wasip1.go
+++ b/src/runtime/lock_wasip1.go
@@ -75,11 +75,13 @@ func notewakeup(n *note) {
 }
 
 func notesleep(n *note) {
-	throw("notesleep not supported by wasi")
+    // throw("notesleep not supported by wasi")
+    // No-op since WASM is single-threaded.
+    return
 }
 
 func notetsleep(n *note, ns int64) bool {
-	throw("notetsleep not supported by wasi")
+	// throw("notetsleep not supported by wasi")
 	return false
 }
 
diff --git a/src/runtime/os_wasip1.go b/src/runtime/os_wasip1.go
index ed4f646a62..70f8801cca 100644
--- a/src/runtime/os_wasip1.go
+++ b/src/runtime/os_wasip1.go
@@ -189,6 +189,7 @@ func usleep(usec uint32) {
 }
 
 func readRandom(r []byte) int {
+	throw("wasip1 readRandom should not be called! Use GORANDSEED to seed runtime")
 	if random_get(&r[0], size(len(r))) != 0 {
 		return 0
 	}
diff --git a/src/runtime/proc.go b/src/runtime/proc.go
index 005c875cbf..39a1ff9a81 100644
--- a/src/runtime/proc.go
+++ b/src/runtime/proc.go
@@ -873,7 +873,7 @@ func schedinit() {
 	ticks.init() // run as early as possible
 	moduledataverify()
 	stackinit()
-	randinit() // must run before mallocinit, alginit, mcommoninit
+	randinit() // must run before mallocinit, AlgInit, mcommoninit
 	mallocinit()
 	cpuinit(godebug) // must run before alginit
 	alginit()        // maps, hash, rand must not be used before this call
@@ -7468,7 +7468,8 @@ func runqempty(pp *p) bool {
 // With the randomness here, as long as the tests pass
 // consistently with -race, they shouldn't have latent scheduling
 // assumptions.
-const randomizeScheduler = raceenabled
+// const randomizeScheduler = raceenabled
+const randomizeScheduler = true
 
 // runqput tries to put g on the local runnable queue.
 // If next is false, runqput adds g to the tail of the runnable queue.
diff --git a/src/runtime/rand.go b/src/runtime/rand.go
index 1739e9f8f5..9ee8ffc32e 100644
--- a/src/runtime/rand.go
+++ b/src/runtime/rand.go
@@ -11,6 +11,7 @@ import (
 	"internal/chacha8rand"
 	"internal/goarch"
 	"internal/runtime/math"
+	"internal/stringslite"
 	"unsafe"
 	_ "unsafe" // for go:linkname
 )
@@ -33,6 +34,33 @@ var globalRand struct {
 
 var readRandomFailed bool
 
+// DST: read user defined seed from env (will only take effect when >= 16 char long!)
+func getGorandseed() {
+	const prefix = "GORANDSEED="
+	switch GOOS {
+	case "aix", "darwin", "ios", "dragonfly", "freebsd", "netbsd", "openbsd", "illumos", "solaris", "linux", "wasip1":
+		// Similar to goenv_unix but extracts the environment value for
+		// GORANDSEED directly.
+		n := int32(0)
+		for argv_index(argv, argc+1+n) != nil {
+			n++
+		}
+
+		for i := int32(0); i < n; i++ {
+			p := argv_index(argv, argc+1+i)
+			s := unsafe.String(p, findnull(p))
+
+			if stringslite.HasPrefix(s, prefix) {
+				if len(startupRand) < 16 {
+					startupRand = make([]byte, 16)
+				}
+				copy(startupRand, s[len(prefix):])
+				break
+			}
+		}
+	}
+}
+
 // randinit initializes the global random state.
 // It must be called before any use of grand.
 func randinit() {
@@ -42,6 +70,10 @@ func randinit() {
 	}
 
 	seed := &globalRand.seed
+
+	// DST: allow user defined seed
+	getGorandseed()
+
 	if len(startupRand) >= 16 &&
 		// Check that at least the first two words of startupRand weren't
 		// cleared by any libc initialization.
