diff --git a/src/race.bash b/src/race.bash
index 3ce91a345f..77fa216613 100755
--- a/src/race.bash
+++ b/src/race.bash
@@ -33,5 +33,5 @@ if [ ! -f make.bash ]; then
 	exit 1
 fi
 . ./make.bash --no-banner
-go install -race std
-go tool dist test -race
+#go install -race std
+#go tool dist test -race
diff --git a/src/run.bash b/src/run.bash
index cb474325c4..e7094aaa3c 100755
--- a/src/run.bash
+++ b/src/run.bash
@@ -50,4 +50,4 @@ if ulimit -T &> /dev/null; then
 fi
 
 export GOPATH=/nonexist-gopath
-exec ../bin/go tool dist test -rebuild "$@"
+# exec ../bin/go tool dist test -rebuild "$@"
diff --git a/src/runtime/lock_wasip1.go b/src/runtime/lock_wasip1.go
index 55153c3a05..55ffab233d 100644
--- a/src/runtime/lock_wasip1.go
+++ b/src/runtime/lock_wasip1.go
@@ -75,11 +75,13 @@ func notewakeup(n *note) {
 }
 
 func notesleep(n *note) {
-	throw("notesleep not supported by wasi")
+    // throw("notesleep not supported by wasi")
+    // No-op since WASM is single-threaded.
+    return
 }
 
 func notetsleep(n *note, ns int64) bool {
-	throw("notetsleep not supported by wasi")
+	// throw("notetsleep not supported by wasi")
 	return false
 }
 
diff --git a/src/runtime/os_wasip1.go b/src/runtime/os_wasip1.go
index ed4f646a62..70f8801cca 100644
--- a/src/runtime/os_wasip1.go
+++ b/src/runtime/os_wasip1.go
@@ -189,6 +189,7 @@ func usleep(usec uint32) {
 }
 
 func readRandom(r []byte) int {
+	throw("wasip1 readRandom should not be called! Use GORANDSEED to seed runtime")
 	if random_get(&r[0], size(len(r))) != 0 {
 		return 0
 	}
diff --git a/src/runtime/proc.go b/src/runtime/proc.go
index 005c875cbf..be65c1ce14 100644
--- a/src/runtime/proc.go
+++ b/src/runtime/proc.go
@@ -873,7 +873,8 @@ func schedinit() {
 	ticks.init() // run as early as possible
 	moduledataverify()
 	stackinit()
-	randinit() // must run before mallocinit, alginit, mcommoninit
+	dstinit()  // must run before randinit
+	randinit() // must run before mallocinit, AlgInit, mcommoninit
 	mallocinit()
 	cpuinit(godebug) // must run before alginit
 	alginit()        // maps, hash, rand must not be used before this call
@@ -7468,7 +7469,8 @@ func runqempty(pp *p) bool {
 // With the randomness here, as long as the tests pass
 // consistently with -race, they shouldn't have latent scheduling
 // assumptions.
-const randomizeScheduler = raceenabled
+// const randomizeScheduler = raceenabled
+const randomizeScheduler = true
 
 // runqput tries to put g on the local runnable queue.
 // If next is false, runqput adds g to the tail of the runnable queue.
diff --git a/src/runtime/dst_linux.go b/src/runtime/dst_linux.go
new file mode 100644
index 0000000000..8fa8feec9d
--- /dev/null
+++ b/src/runtime/dst_linux.go
@@ -0,0 +1,41 @@
+package runtime
+
+import (
+	"internal/stringslite"
+	"unsafe"
+	_ "unsafe" // for go:linkname
+)
+
+// use static allocation because malloc is not yet available
+var dstSeed [16]byte
+
+// DST: read user defined seed from env
+func dstinit() {
+	const prefix = "GORANDSEED="
+	// Similar to goenv_unix but extracts the environment value for
+	// GORANDSEED directly.
+	n := int32(0)
+	for argv_index(argv, argc+1+n) != nil {
+		n++
+	}
+
+	for i := int32(0); i < n; i++ {
+		p := argv_index(argv, argc+1+i)
+		s := unsafe.String(p, findnull(p))
+
+		if stringslite.HasPrefix(s, prefix) {
+			s = s[len(prefix):]
+			if stringslite.HasPrefix(s, "0x") {
+				s = s[2:]
+			}
+			if len(startupRand) < 16 {
+				startupRand = (*[16]byte)(unsafe.Pointer(&dstSeed))[:]
+			}
+			var l int
+			for l < 16 {
+				l += copy(startupRand[l:], s)
+			}
+			break
+		}
+	}
+}
diff --git a/src/runtime/dst_wasip1.go b/src/runtime/dst_wasip1.go
new file mode 100644
index 0000000000..0ff3e0a848
--- /dev/null
+++ b/src/runtime/dst_wasip1.go
@@ -0,0 +1,65 @@
+//go:build wasip1
+
+package runtime
+
+import (
+	"internal/stringslite"
+	"unsafe"
+	_ "unsafe" // for go:linkname
+)
+
+// use static allocation because malloc is not yet available
+var (
+	dstSeed [16]byte
+	dstEnv  [4096]byte
+)
+
+// DST: read user defined seed from env
+func dstinit() {
+	const prefix = "GORANDSEED="
+	// Similar to wasip1:goenvs but extracts the environment value for
+	// GORANDSEED directly.
+
+	// environment
+	var environCount size
+	var environBufLen size
+	if environ_sizes_get(&environCount, &environBufLen) != 0 {
+		throw("environ_sizes_get failed")
+	}
+
+	if 4*environCount+environBufLen > 4096 {
+		throw("environ buffer too small")
+	}
+
+	if environCount > 0 {
+		environ := unsafe.Slice((*uintptr32)(unsafe.Pointer(&dstEnv[0])), environCount*4) // make([]uintptr32, environCount)
+		environBuf := dstEnv[len(environ) : size(len(environ))+environBufLen]             // make([]byte, environBufLen)
+		if environ_get(&environ[0], &environBuf[0]) != 0 {
+			throw("environ_get failed")
+		}
+
+		for i := range environCount {
+			start := environ[i] - uintptr32(uintptr(unsafe.Pointer(&environBuf[0])))
+			end := start
+			for environBuf[end] != 0 {
+				end++
+			}
+			s := unsafe.String(&environBuf[start], end-start)
+
+			if stringslite.HasPrefix(s, prefix) {
+				s = s[len(prefix):]
+				if stringslite.HasPrefix(s, "0x") {
+					s = s[2:]
+				}
+				if len(startupRand) < 16 {
+					startupRand = (*[16]byte)(unsafe.Pointer(&dstSeed))[:]
+				}
+				var l int
+				for l < 16 {
+					l += copy(startupRand[l:], s)
+				}
+				break
+			}
+		}
+	}
+}
